### DISCUSSION TITLE
Is it ok to create nested effect?

### DISCUSSION THREAD

**Discussion by zhaoyao91** [solidjs/solid#379] [Q&A] [answered]
Is it ok to create nested effect?
If it is, what is the expected behavior?

**Comment by ryansolid** [comment, marked as answer]
Yes. And only the nearest parent effect tracks. So if the internal one re-evals it does so as you'd expect. If the external one does it disposes of the internal one, and through re-running recreates it.

**Reply by zhaoyao91** [reply]
I found if I use onCleanup in effect, the cleanup callback will get rerun before each time the effect rerun. Is this by design?

**Reply by ryansolid** [reply]
Yes because presumably what it created previously will be created again. Similar to React's `useEffect` return function. It runs on final cleanup and before each execution.
